Changed tune.report import for ray>=2

drwillcharles opened the pull request:
<!-- Thank you for your contribution! Please review https://microsoft.github.io/FLAML/docs/Contribute before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->
When using ray>=2, there is an import error for `tune.report`.
>AttributeError: module 'ray.tune' has no attribute 'report'

I used the Ray [migration guide](https://docs.google.com/document/d/1o-J5UAeH4M5gGN9BwdrCt1q-nLe58mp94xsmTaW0Flc/edit?tab=t.0) to update this and use the new report function if the ray version is not 1.x .

I see that in the FLAML codebase there are already a few workarounds like this so I think I can do something similar without needing a full migration to Ray 2.x
## Related issue number

<!-- For example: "Closes #1234" -->
Related to https://github.com/microsoft/FLAML/issues/874 - although I still think a full migration to Ray>=2 is needed in the future.
## Checks

- [ x] I've used [pre-commit](https://microsoft.github.io/FLAML/docs/Contribute#pre-commit) to lint the changes in this PR (note the same in integrated in our CI checks).
- [x ] I've included any doc changes needed for https://microsoft.github.io/FLAML/. See https://microsoft.github.io/FLAML/docs/Contribute#documentation to build and test documentation locally.
- [ x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x ] I've made sure all auto checks have passed.

drwillcharles commented:
@microsoft-github-policy-service agree company="Evotec"

thinkall reviewed (commented):
Thank you, @drwillcharles , for the PR!

thinkall commented:
Hi @drwillcharles , could you check the test errors? Looks like the change will lead to test timeout issue. Thanks.

drwillcharles commented:
It appears the test `_test_hf_data()` on `test_gpu.py` gets stuck in a loop because of an error in Ray. Strangely - this test is not even run because of the naming convention. (`_test_hf_data()` instead of `test_hf_data()`).

I have loosened the pydantic dependency and hope this solves the issue.

drwillcharles commented:
[OpenML](www.openml.org) is currently down so a lot of tests are getting connection errors. Will wait until this is back online before I resume development. 

From the tests that have run successfully, it appears there are still some environment issues to fix

drwillcharles commented:
I've figured out that the jobs are timing out because Ray cannot find the appropriate resource 
![image](https://github.com/user-attachments/assets/139acaa6-5727-4c02-be9f-f2756361931f)

There is a function `_test_hf_data` in test/test_gpu.py which I don't think is even picked up in the tests (because of the underscore). 
However, when running this file, it gets stuck in an infinite loop. I think before this just failed silently.

Can I delete this?

thinkall commented:
> I've figured out that the jobs are timing out because Ray cannot find the appropriate resource ![image](https://private-user-images.githubusercontent.com/64903662/407256088-139acaa6-5727-4c02-be9f-f2756361931f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkyNjQwODgsIm5iZiI6MTczOTI2Mzc4OCwicGF0aCI6Ii82NDkwMzY2Mi80MDcyNTYwODgtMTM5YWNhYTYtNTcyNy00YzAyLWJlOWYtZjI3NTYzNjE5MzFmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjExVDA4NDk0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI0YjA1YzU2NmUyYTQzMzEzZTdiNzI2OGIwOThhMzAwMTUwOGMzNDNiZmI0N2ZlZGI2MDI3ZGU5MTUzZTAxYmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.1ZZs6M50PlKagu_Y6zsUTJAT1HgLFIW32C0tKO_RiFo)
> 
> There is a function `_test_hf_data` in test/test_gpu.py which I don't think is even picked up in the tests (because of the underscore). However, when running this file, it gets stuck in an infinite loop. I think before this just failed silently.
> 
> Can I delete this?

Sorry for the late reply, @drwillcharles. It's interesting that this underscored test is picked up. For bypassing the test, I think we can  detect whether there is a GPU resource and skip the test if None. WDYT? Thanks.

drwillcharles commented:
No problem @thinkall. I did add a check for the GPU resource and even deleted the test to check. Turns out I was wrong and the problem appears to be `test_lrl2` on test_model.py.

Looking at the line `lr.fit(X, y, budget=1e-5)` it appears the test should just stop when the budget is exceeded (almost instantaneous). However, that appears not to be happening. I have not yet found a solution to this and it appears https://github.com/microsoft/FLAML/pull/1402 is having the same problem in the tests.

thinkall commented:
> No problem @thinkall. I did add a check for the GPU resource and even deleted the test to check. Turns out I was wrong and the problem appears to be `test_lrl2` on test_model.py.
> 
> Looking at the line [`lr.fit(X, y, budget=1e-5)`](https://github.com/microsoft/FLAML/blob/6d539298036b144b968dd70b1089fb9436eef4b1/test/test_model.py#L26) it appears the test should just stop when the budget is exceeded (almost instantaneous). However, that appears not to be happening. I have not yet found a solution to this and it appears #1402 is having the same problem in the tests.

Thank you for pointing out the root cause test, @drwillcharles . I fixed the tests in #1406 .

drwillcharles commented:
> > No problem @thinkall. I did add a check for the GPU resource and even deleted the test to check. Turns out I was wrong and the problem appears to be `test_lrl2` on test_model.py.
> > Looking at the line [`lr.fit(X, y, budget=1e-5)`](https://github.com/microsoft/FLAML/blob/6d539298036b144b968dd70b1089fb9436eef4b1/test/test_model.py#L26) it appears the test should just stop when the budget is exceeded (almost instantaneous). However, that appears not to be happening. I have not yet found a solution to this and it appears #1402 is having the same problem in the tests.
> 
> Thank you for pointing out the root cause test, @drwillcharles . I fixed the tests in #1406 .

Thanks @thinkall ! I have implemented your fix so hopefully these tests pass now 🤞
